pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -730,13 +729,47 @@ public void createConsumingSegment(String realtimeTableName, List<Integer> nonCo
           int nextSeqNum = segmentName.getSequenceNumber() + 1;
           List<String> instances = partitionAssignment.getListField(Integer.toString(curPartition));
           LOGGER.info("Creating CONSUMING segment for {} partition {} with seq {}", realtimeTableName, curPartition, nextSeqNum);
+          // To begin with, set startOffset to the oldest availble offset in kafka. Fix it to be the one we want,
+          // depending on what the prev segment had.
           long startOffset = getKafkaPartitionOffset(kafkaStreamMetadata, "smallest", curPartition);
           LOGGER.info("Found kafka offset {} for table {} for partition {}", startOffset, realtimeTableName, curPartition);
+
           final LLCRealtimeSegmentZKMetadata oldSegMetadata = getRealtimeSegmentZKMetadata(realtimeTableName,
               segmentName.getSegmentName());
-          if (startOffset < oldSegMetadata.getEndOffset()) {
-            startOffset = oldSegMetadata.getEndOffset();
-            LOGGER.info("Choosing newer kafka offset {} for table {} for partition {}", startOffset, realtimeTableName, curPartition);
+          CommonConstants.Segment.Realtime.Status status = oldSegMetadata.getStatus();
+          final long prevSegStartOffset = oldSegMetadata.getStartOffset();  // Offset at which the prev segment intended to start consuming
+          if (status.equals(CommonConstants.Segment.Realtime.Status.IN_PROGRESS)) {
```

Comment:
Yes, let us discuss on how we can make the core more readable. 

Meanwhile, the loop does not fix any missing (kafka) partitions. It only adds segments to a partition if needed.

If the partition is consuming fine, then (no segments need to be created), it is handled between lines 711 and 715.

If the partition has no segments at all to begin with, then (we need to create the first segment with earliest offset) that is handled in lines 717 - 725.

If  a partition already has segments, and the highest segment in that partition is not consuming, then (we need to create a new segment with appropriate offset) that is handled in lines 729 to 777.